### PR TITLE
feat: Implement `App` API

### PR DESCRIPTION
**Pull request by bnjbvr**
This implements a very high-level `App` API that takes care of all the sliding syncs:

- if configured with `with_encryption_sync`, then it will spawn two sliding syncs, one for the room list, one for e2ee. It can be configured to run with the cross-process crypto store lock or not (for the needs of Element X iOS).
- if not, it will run a single sliding sync that powers both the room list and e2ee.

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/1928.

This is based on top of #2235, as the final commits get rid of ways to manually create (and misconfigure) `EncryptionSync` and the `RoomListService` \o/

### CODE REVIEW

File: crates/matrix-sdk-ui/src/app/mod.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            identifier: "app".to_owned(),
+        }
+    }
+
+    pub fn identifier(mut self, name: String) -> Self {
+        self.identifier = name;
+        self
+    }
+
+    pub fn with_encryption_sync(mut self) -> Self {
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/app/mod.rs:163]: As mentioned in the Matrix room, do we need this? Can't we make it always true?

File: crates/matrix-sdk-ui/src/app/mod.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+#[derive(Clone)]
+pub enum AppState {
+    Running,
+    Terminated,
+    Error,
+}
+
+pub struct App {
+    room_list: Arc<RoomListService>,
+    encryption_sync: Option<Arc<EncryptionSync>>,
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/app/mod.rs:42]: If we're always going to have an `EncryptionSync` then we don't need an optional here. We could `#[cfg(feature = "e2e-encryption")]` this field if we want to support clients that don't use E2EE.

File: crates/matrix-sdk-ui/src/app/mod.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+            if let Some(encryption_sync) = encryption_sync {
+                let encryption_sync_stream = encryption_sync.sync();
+
+                pin_mut!(encryption_sync_stream);
+
+                // Note: any error on one of the two syncs will cause the overall stream to error
+                // and the loop to terminate.
+
+                loop {
+                    tokio::select! {
```
Comment (poljar) [inline crates/matrix-sdk-ui/src/app/mod.rs:88]: Nice use of a `select!`.

**Review by poljar — CHANGES_REQUESTED** [review]
Very nice, I think all of it makes sense. If we can remove the ability to disable E2EE on the main client, that would be even better.

File: crates/matrix-sdk-ui/src/app/mod.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/app/mod.rs:163]: I think that's needed until we enable the encryption sync by default... which should be the case any time now.

Comment (poljar) [reply crates/matrix-sdk-ui/src/app/mod.rs:163]: Alright, once we do that, please let us remove this as well.

**Comment by Velin92** [comment]
Tested and works pretty well on the iOS side, makes the code easier to read a decouples a lot of responsibilities from the client, love it!

**Comment by bnjbvr** [comment]
Thanks a bunch Mauro for trying it :partying_face: 

This is now ready for review, since the `NotificationClient` API has been merged. I haven't included any proper tests for App here, since they would mostly repeat what the room list and/or encryption sync do, but I could add some if you have any ideas for specific tests.

I still am not fond of the `App` name: IMO it's too generic, and doesn't quite describe what this does. How about any of the following: `SyncService` / `SyncLoops` / `Synchronizer` (lol) / `SyncLoopService`? (or anything else that tells it's doing some syncs)

**Review by poljar — APPROVED** [review]
Looks good. Not a fan of the `App` name either, but we can bikeshed later on.

**Pull request merged**